### ISSUE TITLE
Another _Ember is undefined error

### ISSUE THREAD

**Issue by sandydoo** [babel/ember-cli-babel#346]
#### Symptoms

Same symptoms as #342, but the errors are coming up while *loading tests in Travis*. Every `require` fails with this error.

#### How to replicate

~~You should be able to replicate by running tests with `CI=true`.~~

It's related to this default line in `targets.js`. Removing `ie 11` as a target removes the issue.

```js
if (isCI || isProduction) {
  browsers.push('ie 11');
}
```

Tested on latest `ember-cli-babel@7.20.4`.

**Comment by sandydoo** [comment]
Okay, so I've managed to track down and fix my issues with some much needed refactoring.
Granted, what "previous me" was doing was dumb 🤣, but this 'position' nonsense below seems to have confused the transpiler. I'll leave this up in case it helps improve anything here.

Original:
```js
import { computed, getProperties } from '@ember/object';

function position() {
  return computed('lat', 'lng', function() {
    const { lat, lng } = getProperties(this, 'lat', 'lng');
    return (lat && lng) ? new google.maps.LatLng(lat, lng) : undefined;
  });
}

export { position };
```
Compiled:
```js
define("ember-google-maps/utils/helpers", ["exports"], function (_exports) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.position = position;

  function position() {
    return (_Ember = Ember).computed.apply('lat', 'lng', function () {
      var _Ember$getProperties = Ember.getProperties(this, 'lat', 'lng'),
          lat = _Ember$getProperties.lat,
          lng = _Ember$getProperties.lng;

      return lat && lng ? new google.maps.LatLng(lat, lng) : undefined;
    });
  }
});
```

One last thing! This similar function below compiles correctly. I guess the transpiler is catching on to the `...args`.

```js
function computedPromise(...args) {
  let func = args.pop();
  return computed(...args, function() {
    return ObjectPromiseProxy.create({
      promise: func.apply(this)
    });
  });
}
```

**Comment by rwjblue** [comment]
> loading tests in Travis

Which repo was this for? I’d like to repro locally to see what can be changed in the Babel plugin.

**Comment by BarryThePenguin** [comment]
I'm seeing a similar thing, running `CI=true ember serve` with the following component

```js
import Component from '@ember/component'
import { computed, set } from '@ember/object'

import layout from './template'

function setSomeComputedKeys(keys) {
  const someDataCP = computed(...keys, this.someDataCPFunction)
  set(this, 'someData', someDataCP)
}

function someDataCPFunction() {
  return []
}

export default Component.extend({
  layout,
  tagName: '',

  init(...args) {
    this._super(...args)
    setSomeComputedKeys.call(this, [])
  },

  someDataCPFunction,

  data: computed('someData.[]', function () {
    return []
  }),
})
```

Compiles to

```js
;define("dummy/components/foo/component", ["exports", "dummy/components/foo/template"], function (_exports, _template) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.default = void 0;

  function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread(); }

  function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

  function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

  function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && Symbol.iterator in Object(iter)) return Array.from(iter); }

  function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) return _arrayLikeToArray(arr); }

  function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

  function setSomeComputedKeys(keys) {
    var _Ember;

    var someDataCP = (_Ember = Ember).computed.apply(_Ember, _toConsumableArray(keys).concat([this.someDataCPFunction]));

    Ember.set(this, 'someData', someDataCP);
  }

  function someDataCPFunction() {
    return [];
  }

  var _default = Ember.Component.extend({
    layout: _template.default,
    tagName: '',
    init: function init() {
      this._super(arguments);

      setSomeComputedKeys.call(this, []);
    },
    someDataCPFunction: someDataCPFunction,
    data: (_Ember = Ember).computed.apply('someData.[]', function () {
      return [];
    })
  });

  _exports.default = _default;
});
```

For some reason, removing the usage of the spread operator (`...`) from the component appears to fix the issue.

**Comment by sandydoo** [comment]
> Which repo was this for? I’d like to repro locally to see what can be changed in the Babel plugin.

This was in ember-google-maps. I've created a branch out of the old version for you. No setup necessary, will error out right away.

https://github.com/sandydoo/ember-google-maps/tree/ember-cli-babel-repro

**Comment by rwjblue** [comment]
Awesome, thank you @sandydoo!

**Comment by devdemi** [comment]
I have the same problem as @sandydoo mentioned. 
In my `package.json` I have ` "ember-cli-babel": "~7.20.4"` but I called `yarn why ember-cli-babel` and a lot of libraries depends on `ember-cli-babel: 6.18.0`, e.g. I use `ember-cli-notifications: ^4.3.3` which depends on `"ember-css-modules": "^0.7.10"` which depends on `ember-cli-babel: 6.18.0`, seems two version of `ember-cli-babel` conflicts each other.

**Comment by rwjblue** [comment]
> seems two version of ember-cli-babel conflicts each other

No, they should not conflict with each other. Each addon will use the version of ember-cli-babel that they directly depend on to transpile their own code.

**Comment by devdemi** [comment]
@rwjblue if I turn on 
```
if (isCI || isProduction) {
  browsers.push('ie 11');
}
```
I have the next error:
```
ReferenceError: _Ember is not defined
    at configurable (/usr/src/fastboot/dist/assets/addon-tree-output/torii/configuration.js:13:1)```

```
I use `torii: 0.9.6` which depends on `"ember-cli-babel": "^6.3.0"`. But I have a lot of libraries which depend on `"ember-cli-babel": "^6.3.0"`

**Comment by rwjblue** [comment]
OK, that sounds a lot like the scenario that @sandydoo called out and provided a reproduction for. I'll be looking into it this afternoon.

**Comment by cah-briangantzler** [comment]
Same problem here with ember-metrics (just posted in the discord). Not sure what Babel was trying to do with `(_Ember = Ember)` but it looks like its going to be happening in lots of places in various addons. Assuming its something in the 7.20.x release

**Comment by richard-viney** [comment]
Just a note that the build for the PR at ember-intl/ember-intl#1340 is encountering this issue as well. It's using `ember-cli-babel@7.20.4`, and the issue reproduces locally as well.

The Travis run is here: https://travis-ci.com/github/ember-intl/ember-intl/jobs/342607182

**Comment by rwwagner90** [comment]
@rwjblue all of my addons, ember-shepherd. ember-math-helpers, ember-flatpickr, etc are experiencing this issue. I am unable to run any CI builds. Is there a workaround?

**Comment by ananddotmurali** [comment]
We too had this issue. Fixed it with pinning "babel-plugin-ember-modules-api-polyfill": "2.13.0". Kind of workaround but works.

**Comment by ananddotmurali** [comment]
We too had this issue. Fixed it with pinning "babel-plugin-ember-modules-api-polyfill": "2.13.0". Kind of workaround but works.

**Comment by cah-briangantzler** [comment]
@ananddotmurali could you supply details on what you did to pin that, as what I tried did not work

**Comment by sandydoo** [comment]
> I am unable to run any CI builds. Is there a workaround?

@rwwagner90, a few workarounds we know of.

1. Stay on `ember-cli-babel@7.19.*` for now. For CI builds with floating dependencies that means pinning down the version.

2. Remove the `ie*` target from `targets.js`. This is an ember-cli default, removing it will bypass the issue.

**Comment by sandydoo** [comment]
> ember-math-helpers, ember-flatpickr, etc are experiencing this issue

So I was curious... Why would ember-math-helpers encounter this error? It doesn't even have any computed properties. Turns out it's because of ember-cli-addon-docs.

https://github.com/ember-learn/ember-cli-addon-docs/blob/f1089775f32b56a61392b62eee36d26e06a68983/addon/utils/computed.js#L8-L29

**Comment by richard-viney** [comment]
Interesting, my repro for this issue in ember-intl/ember-intl#1340 occurs in the exact same place inside `ember-cli-adon-docs`.

**Comment by sandydoo** [comment]
I think we've got a fix? Go @rwjblue🥇 

https://github.com/ember-cli/babel-plugin-ember-modules-api-polyfill/commit/621360cb8ccc08d0c36f5bfe4f44548fce9648f9

**Comment by rwjblue** [comment]
Ya, sorry for this taking so long. It took a little while to figure out which specific babel plugins were mutating the nodes, and I only discovered that using the shared AST nodes was causing an issue due to stepping through the tranpsilation process.

#348 should land and be released as 7.20.5 soon.

**Comment by rwjblue** [comment]
https://github.com/babel/ember-cli-babel/releases/tag/v7.20.5 is published now with the updates from https://github.com/ember-cli/babel-plugin-ember-modules-api-polyfill/pull/109. I tried to explain exactly what was going on in #348 and https://github.com/ember-cli/babel-plugin-ember-modules-api-polyfill/pull/109.

Note: the actual fix was in https://github.com/ember-cli/babel-plugin-ember-modules-api-polyfill/pull/109, so you can almost certainly get these fixes _without_ being on ember-cli-babel@7.20.x, but in order to do so you'd need to ensure your lockfile doesn't keep the babel-plugin-ember-modules-api-polyfill pinned. 

For example, for `yarn` users you can do:

* add a resolution for `babel-plugin-ember-modules-api-polyfill` to use `2.13.4`
* run `yarn`
* remove the resolution
* run `yarn` again
